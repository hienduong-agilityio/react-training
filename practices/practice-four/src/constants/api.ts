export const HTTP_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE'
};

export const ROOT_API = import.meta.env.VITE_API_ENDPOINT;

export const API_URL = {
  PROJECT: `${ROOT_API}/projects`
};
