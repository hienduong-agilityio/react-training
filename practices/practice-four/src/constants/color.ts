// Consider using enums instead of constants
export const COLORS_CLASS = {
  DEFAULT: 'color-default',
  PRIMARY: 'color-primary',
  SUCCESS: 'color-success',
  WARNING: 'color-warning',
  DANGER: 'color-danger '
};
