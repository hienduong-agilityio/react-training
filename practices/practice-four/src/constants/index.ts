// Index constants files
export * from './api';
export * from './color';
export * from './filerSearchOption';
export * from './route';
export * from './rowPerPage';
export * from './store';
export * from './validationMessages';
