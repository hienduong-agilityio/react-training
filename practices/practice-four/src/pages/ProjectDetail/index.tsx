// Libraries
import { Navigate, useParams } from 'react-router-dom';

// Components
import { ProjectDetailSkeleton } from '@/components';
import { ProjectDetailInfo } from '@/pages/ProjectDetail/ProjectDetailInfo';

// Hooks
import { useProject } from '@/hooks';

// Interfaces
import { IProjectItemProps } from '@/interfaces';

/**
 * Component to render detailed information for a single project.
 *
 * @returns {JSX.Element} The rendered project details.
 */
const ProjectDetail = (): JSX.Element => {
  const { id } = useParams<{ id: string }>();

  // Get project detail by useProject hook
  const project = useProject({ id: id || '' });
  const { data: projectDetail, isQueryProjectDetailPending, error: projectDetailsError } = project;

  // Render loading spinner if data is being fetched
  if (isQueryProjectDetailPending) {
    return <ProjectDetailSkeleton />;
  }

  // If the project is not found, redirect to a 404 page
  if (!projectDetail) {
    return <Navigate to='/404' replace />;
  }

  // Render error message if there is an error in data fetching
  if (projectDetailsError) {
    return <div>Error loading projects: {projectDetailsError.message}</div>;
  }

  const { manager: { managerName = 'Unknown', managerImage = '' } = {} } = projectDetail as IProjectItemProps;

  return <ProjectDetailInfo managerName={managerName} managerAvatar={managerImage} {...projectDetail} />;
};

export default ProjectDetail;
