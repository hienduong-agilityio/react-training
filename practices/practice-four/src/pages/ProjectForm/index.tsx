'use client';

// Libraries
import { useFormState } from 'react-dom';
import { Navigate, useNavigate } from 'react-router-dom';

// Hooks
import { initialProjectFormState, useProjectForm } from '@/hooks';

// Components
import { FormLayout, ProjectFormFields } from '@/components';

const ProjectFormPage = ({ title = '' }: { title: string }) => {
  const navigate = useNavigate();

  const { formAction, initialFormValues, isQueryProjectDetailPending, projectDetailsError } = useProjectForm(title);

  /**
   * useFormState is a hook that allows you to manage the state of a form.
   */
  const [formState, formActionDispatcher] = useFormState(formAction, initialProjectFormState);

  if (projectDetailsError) return <Navigate to='/404' replace />;

  const handleCancelForm = () => {
    if (window.history.length > 1) {
      navigate(-1);
    }
  };
  return (
    <FormLayout title={title}>
      <ProjectFormFields
        initialFormValues={initialFormValues}
        formErrorsMessages={formState.errors}
        isQueryProjectDetailPending={isQueryProjectDetailPending}
        title={title}
        handleCancelForm={handleCancelForm}
        formAction={formActionDispatcher}
      />
    </FormLayout>
  );
};

export default ProjectFormPage;
