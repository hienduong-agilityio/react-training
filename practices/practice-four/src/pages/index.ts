// Index Page files
export { default as NotFoundPage } from './NotFound';
export { default as ProjectDetail } from './ProjectDetail';
export { default as ProjectFormPage } from './ProjectForm';
export { default as ProjectPage } from './ProjectTable';
