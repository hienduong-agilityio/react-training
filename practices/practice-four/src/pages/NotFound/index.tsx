// Libraries
import { Link } from 'react-router-dom';
import { usePageSeo } from '@/hooks';

// Constants
import { ROUTE } from '@/constants';

/**
 * NotFoundPage
 *
 * @returns {JSX.Element} - The NotFound element
 */
const NotFoundPage = (): JSX.Element => {
  usePageSeo({
    title: 'Page Not Found',
    description: 'The page you are looking for does not exist.',
    keywords: '404, not found, error',
    ogTitle: 'Page Not Found',
    ogDescription: 'The page you are looking for does not exist.'
  });

  return (
    <div className='flex flex-col h-screen justify-center items-center bg-gray-300'>
      <p className='text-[120px] font-extrabold text-gray-700'>404</p>
      <p className='text-2xl font-medium text-gray-600 mb-6'>Not Found</p>
      <Link
        to={ROUTE.ROOT}
        className='px-4 py-2 font-medium text-white bg-primary-500 rounded-md hover:bg-primary-600 transition-all duration-200 ease-in-out'
      >
        Go Home
      </Link>
    </div>
  );
};

export default NotFoundPage;
