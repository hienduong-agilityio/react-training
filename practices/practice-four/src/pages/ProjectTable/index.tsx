import { Suspense, useDeferredValue, useOptimistic, startTransition, useActionState, useState } from 'react';
import { useNavigate, useLocation, useSearchParams, Outlet } from 'react-router-dom';
import { useQuery, keepPreviousData } from '@tanstack/react-query';
import { useScan, Options } from 'react-scan';

// Components
import { ProjectSearch, Button, Spinner, ProjectTableManager } from '@/components';

// Constants
import { ROUTE, ROWS_PER_PAGE_OPTIONS } from '@/constants';

// Hooks
import { usePageSeo, useProject, useSearchProject } from '@/hooks';

// Services
import { getProjects } from '@/services';

// Types
import type { IProjectsQueryResult } from '@/interfaces';

const ProjectPage = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const [searchParams] = useSearchParams();

  usePageSeo({
    title: 'Project Management',
    description: '"A powerful tool for managing and tracking projects efficiently.',
    keywords: 'project, management, tool',
    ogTitle: 'Project Management',
    ogDescription: '"A powerful tool for managing and tracking projects efficiently.',
    ogImage: 'https://example.com/og-image.jpg',
    ogUrl: 'https://project-managerment-table.netlify.app/'
  });

  const isChildRoute = location.state?.from;

  // Get values from URL
  const currentPage = parseInt(searchParams.get('page') ?? '1');
  const rowsPerPage = parseInt(searchParams.get('rowsPerPage') ?? '10');
  const initialField = searchParams.get('searchField') ?? 'projectName';
  const initialKeyword = searchParams.get(initialField) ?? '';

  // Search state hook
  const { searchField, searchKeyword, updateSearchTerm, updateSearchField } = useSearchProject(
    initialField,
    initialKeyword
  );

  const searchAction = async (_prevState: { searchValue: string }, formData: FormData) => {
    const keyword = formData.get('search') as string;

    updateSearchTerm(keyword);

    return { searchValue: keyword };
  };

  const searchFormStateTuple = useActionState(searchAction, {
    searchValue: initialKeyword
  });

  const formActionDispatcher = searchFormStateTuple[1];

  const [optimisticState, updateOptimistic] = useOptimistic({ currentPage, rowsPerPage }, (prev, next) => ({
    ...(prev ?? {}),
    ...(next ?? {})
  }));

  const deferredKeyword = useDeferredValue(searchKeyword);

  const {
    data: projectData,
    isQueryProjectsPending,
    error
  } = useProject({
    page: optimisticState.currentPage,
    rowsPerPage: optimisticState.rowsPerPage,
    filter: { [searchField]: deferredKeyword }
  });

  const allProjectsQuery: IProjectsQueryResult = useQuery({
    queryKey: ['projects', searchField, deferredKeyword],
    queryFn: () => getProjects({ filter: { [searchField]: deferredKeyword } }),
    placeholderData: keepPreviousData
  });

  const totalPages = Math.ceil((allProjectsQuery.data?.length ?? 1) / optimisticState.rowsPerPage);

  useScan({
    enabled: import.meta.env.DEV,
    name: 'ProjectPage',
    logProps: true,
    includeComponents: ['ProjectSearch', 'ProjectTableManager', 'Button'],
    rowsPerPage: optimisticState.rowsPerPage
  } as Options & { rowsPerPage: number });

  // Pagination handlers
  const handleRowsPerPageChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const newRowsPerPage = parseInt(event.target.value, 10);

    startTransition(() => {
      updateOptimistic({ rowsPerPage: newRowsPerPage, currentPage: 1 });
    });

    const newParams = new URLSearchParams(searchParams.toString());

    newParams.set('rowsPerPage', newRowsPerPage.toString());
    newParams.set('page', '1');

    navigate(`?${newParams.toString()}`);
  };

  const handlePageChange = (newPage: number) => {
    startTransition(() => {
      updateOptimistic({ currentPage: newPage });
    });

    const newParams = new URLSearchParams(searchParams.toString());
    newParams.set('page', newPage.toString());

    navigate(`?${newParams.toString()}`);
  };

  // Delete modal state
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
  const [selectedProjectId, setSelectedProjectId] = useState('');

  const handleOpenDeleteProjectModal = (id: string) => {
    startTransition(() => {
      setSelectedProjectId(id);
      setIsDeleteModalOpen(true);
    });
  };

  const handleDeleteModalClose = () => {
    startTransition(() => {
      setIsDeleteModalOpen(false);
    });
  };

  // Option builder
  const rowPerPageOption = ROWS_PER_PAGE_OPTIONS.map((option) => (
    <option key={option.value} value={option.value}>
      {option.label}
    </option>
  ));

  const handleNewProjectClick = () => {
    navigate(ROUTE.ADD_PROJECT, { state: { from: location.pathname + location.search } });
  };

  const isPending = isQueryProjectsPending || allProjectsQuery.isFetching;

  if (error ?? allProjectsQuery.error) return <div>Error loading projects</div>;

  return (
    <div className='h-full'>
      {!isChildRoute ? (
        <>
          <div className='flex justify-between pt-5 pb-7 px-3'>
            <ProjectSearch
              searchField={searchField}
              updateSearchField={updateSearchField}
              initialKeyword={initialKeyword}
              formAction={formActionDispatcher}
            />
            <Button onClick={handleNewProjectClick}>New Project</Button>
          </div>

          <Suspense fallback={<Spinner />}>
            <ProjectTableManager
              isPending={isPending}
              allProjectsQuery={allProjectsQuery}
              projectData={Array.isArray(projectData) ? projectData : []}
              rowsPerPage={optimisticState.rowsPerPage}
              totalPages={totalPages}
              currentPage={optimisticState.currentPage}
              selectedProjectId={selectedProjectId}
              isDeleteModalOpen={isDeleteModalOpen}
              rowPerPageOption={rowPerPageOption}
              handleRowsPerPageChange={handleRowsPerPageChange}
              handlePageChange={handlePageChange}
              handleOpenDeleteProjectModal={handleOpenDeleteProjectModal}
              handleDeleteModalClose={handleDeleteModalClose}
            />
          </Suspense>
        </>
      ) : (
        <Outlet />
      )}
    </div>
  );
};

export default ProjectPage;
