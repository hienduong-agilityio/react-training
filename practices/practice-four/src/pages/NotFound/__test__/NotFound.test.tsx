// Libraries
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

// Pages
import { NotFoundPage } from '@/pages';

// Constants
import { ROUTE } from '@/constants';

describe('NotFoundPage', () => {
  let component: ReturnType<typeof render>;

  beforeEach(() => {
    component = render(
      <MemoryRouter>
        <NotFoundPage />
      </MemoryRouter>
    );
  });

  it('should match snapshot', () => {
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render 404 text, Not Found message, and Go Home link', () => {
    expect(screen.getByText('404')).toBeInTheDocument();
    expect(screen.getByText('Not Found')).toBeInTheDocument();

    const homeLink = screen.getByRole('link', { name: /Go Home/i });
    expect(homeLink).toBeInTheDocument();
    expect(homeLink).toHaveAttribute('href', ROUTE.ROOT);
  });

  it('should navigate to home when "Go Home" is clicked', async () => {
    const user = userEvent.setup();

    const homeLink = screen.getByRole('link', { name: /Go Home/i });
    await user.click(homeLink);

    expect(homeLink).toHaveAttribute('href', ROUTE.ROOT);
  });
});
