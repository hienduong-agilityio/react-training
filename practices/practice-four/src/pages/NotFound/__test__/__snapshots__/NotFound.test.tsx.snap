// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFoundPage should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col h-screen justify-center items-center bg-gray-300"
  >
    <p
      class="text-[120px] font-extrabold text-gray-700"
    >
      404
    </p>
    <p
      class="text-2xl font-medium text-gray-600 mb-6"
    >
      Not Found
    </p>
    <a
      class="px-4 py-2 font-medium text-white bg-primary-500 rounded-md hover:bg-primary-600 transition-all duration-200 ease-in-out"
      href="/"
    >
      Go Home
    </a>
  </div>
</DocumentFragment>
`;
